New Python `venv` recommendations (#19)
UI-Research/reproducibility-at-urban: virtual-environments.qmd

There is a plethora of virtual environment managers in Python, with a few of the most common being `virtualenv`, `pipenv`, `poetry`, and `conda`.

> Due to changes in the Anaconda terms of services, we are recommending against using conda for virtual environments. `venv` is a great alternative built into base Python that is straightforward to use and has compatibility with `pip`. `virtualenv`, `pipenv` and `poetry` are also popular options worth consideration that may have additional functionality for specific needs. If you're using R and Python, `renv` is a good option given its emphasis on interoperability. Documentation for using Python with `renv` is available <a href="https://rstudio.github.io/renv/articles/python.html" target="_blank">here</a>.

Reviewer: ```suggestion
Due to Anaconda's terms of services, we are recommending against using `conda` for virtual environments. `venv` is a great alternative built into base Python that is straightforward to use and has compatibility with `pip`. `virtualenv`, `pipenv` and `poetry` are also popular options worth consideration that may have additional functionality for specific needs. If you're using R and Python, `renv` is a good option given its emphasis on interoperability. Documentation for using Python with `renv` is available <a href="https://rstudio.github.io/renv/articles/python.html" target="_blank">here</a>.
```

### Python

To set up a virtual environment for a Python project, we recommend using `venv`, which is built into base Python. Note that this may not work with older versions of Python (pre-3.3), in which case you should use [`virtualenv`](https://virtualenv.pypa.io/en/latest/).

::: callout-note
## Where do I run these commands?

All commands below should be run from the Command Prompt or a terminal window (e.g. within RStudio or VS Code) from the root of your project's directory.
:::

1.  **Initialize** a new environment (optionally specifying a version of Python and/or a list of packages) using one of the options below. You should replace `my_env` with a descriptive name specific to your project. 

    ``` python
    python -m venv my_env                                # create empty environment  
    python3.12 -m venv my_env                            # create environment with specific Python version 
    ```

> This will install a directory with your environment's name at the root of your project folder. Unlike `renv`, we don't want to push this folder to GitHub for reproducibility, so make sure to add it to your .gitignore if not already there.

Reviewer: ```suggestion
This will install a directory with your environment's name at the root of your project folder. Unlike `renv`, we don't want to push this folder to GitHub for reproducibility, so make sure to add it to your `.gitignore` if not already there.
```

### Python

We recommend the official `venv` documentation [here](https://docs.python.org/3/tutorial/venv.html).

> In addition to the commands included above, a few other commonly used `conda` commands include:

Reviewer 2: remove mention of conda here?

-   `deactivate` deactivates the current environment
-   `python -m pip list` lists all packages in the current environment
-   `python -m pip show package-name` gives specific information about an installed package
-   Deleting the directory containing the virtual environment will remove the environment from your computer

> 

Author: ```suggestion
See this [Carpentries guide](https://carpentries-incubator.github.io/python-intermediate-development/12-virtual-environments/index.html) for an additional helpful resource
```

-   `deactivate` deactivates the current environment
-   `python -m pip list` lists all packages in the current environment
-   `python -m pip show package-name` gives specific information about an installed package
-   Deleting the directory containing the virtual environment will remove the environment from your computer

> The <a href="https://docs.conda.io/projects/conda/en/latest/user-guide/tasks/manage-environments.html" target="_blank">official documentation</a> and <a href="https://carpentries-incubator.github.io/introduction-to-conda-for-data-scientists/04-sharing-environments/index.html" target="_blank">this guide</a> from the Carpentries are also helpful resources.

Reviewer: Perhaps replace this with the Carpentries resource [here](https://carpentries-incubator.github.io/python-intermediate-development/12-virtual-environments/index.html)? I think Manu (and perhaps Silke or Evy?) tried this out recently and might have other good resources to point folks to.
